File: agent/sap_nexus_agent/execution_result.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class ValidationResult:
    trace_id: str
    capability_id: str
    success: bool
    error_type: str
    messages: list[str]

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "ValidationResult":
        return cls(
            trace_id=str(payload.get("traceId", "")),
            capability_id=str(payload.get("capabilityId", "")),
            success=bool(payload.get("success", False)),
            error_type=str(payload.get("errorType", "UNKNOWN")),
            messages=[str(message) for message in payload.get("messages", [])],
        )


@dataclass(frozen=True)
class ExecutionResult:
    trace_id: str
    capability_id: str
    success: bool
    executor: dict[str, Any]
    return_messages: list[dict[str, Any]]
    data: dict[str, Any]
    duration_ms: int
    error_type: str

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "ExecutionResult":
        data = dict(payload.get("data") or {})
        # Absorb write-path top-level fields into data for unified access.
        # The Gateway WRITE response returns prNumber/commitStatus at the top
        # level (ActionResult shape); read paths keep using nested data.
        if "prNumber" in payload:
            data["prNumber"] = payload["prNumber"]
        if "commitStatus" in payload:
            data["commitStatus"] = payload["commitStatus"]
        return cls(
            trace_id=str(payload.get("traceId", "")),
            capability_id=str(payload.get("capabilityId", "")),
            success=bool(payload.get("success", False)),
            executor=dict(payload.get("executor") or {}),
            return_messages=[dict(message) for message in payload.get("returnMessages", [])],
            data=data,
            duration_ms=int(payload.get("durationMs", 0)),
            error_type=str(payload.get("errorType", "UNKNOWN")),
        )

File: services/gateway/core/src/main/java/com/sapnexus/gateway/result/ExecutionResult.java
package com.sapnexus.gateway.result;

import java.util.List;
import java.util.Map;

public record ExecutionResult(
        String traceId,
        String capabilityId,
        boolean success,
        ExecutorMetadata executor,
        List<SapReturnMessage> returnMessages,
        Map<String, Object> data,
        long durationMs,
        ErrorType errorType
) {
    public static ExecutionResult success(
            String traceId,
            String capabilityId,
            String executorType,
            String rfcName,
            List<SapReturnMessage> returnMessages,
            Map<String, Object> data,
            long durationMs
    ) {
        return new ExecutionResult(
                traceId,
                capabilityId,
                true,
                new ExecutorMetadata(executorType, rfcName),
                returnMessages,
                data,
                durationMs,
                ErrorType.NONE
        );
    }

    public static ExecutionResult failure(
            String traceId,
            String capabilityId,
            String executorType,
            String rfcName,
            ErrorType errorType,
            String message,
            long durationMs
    ) {
        return new ExecutionResult(
                traceId,
                capabilityId,
                false,
                new ExecutorMetadata(executorType, rfcName),
                List.of(new SapReturnMessage("E", "", "", message, "")),
                Map.of(),
                durationMs,
                errorType
        );
    }

    public record ExecutorMetadata(String type, String rfcName) {
    }
}
